### ISSUE TITLE
Expose primitive ordered pixel shaders

### ISSUE THREAD

**Issue by Degerz** [GPUOpen-Drivers/AMDVLK#108]
According to the Vega ISA documentation, this feature uses the SOPP scalar microcode format. Currently, this feature is only exposed in AMD's D3D12 drivers as "rasterizer ordered views" so I'd like to see a mirror equivalent supported in Vulkan as well known as `VK_EXT_fragment_shader_interlock`. 

We need this feature to emulate a certain PowerVR GPU for our use case and particularly we want the `fragmentShaderPixelInterlock` feature exposed from the extension so can your team enable this for the Vulkan drivers ? (Bonus if the team can also get `fragmentShaderSampleInterlock` exposed too) Also if you are working on this extension can we also get an estimate of when a driver will be released to support the extension/feature ?

**Comment by Degerz** [comment]
Can I get a response from the team ?

**Comment by jinjianrong** [comment]
Our stance is that we don’t want to implement it. It messes with subpasses and is not the right formulation

**Comment by Degerz** [comment]
What exactly do you mean by "not the right formulation" ? Is this extension somehow the wrong abstraction to map to this feature inside the hardware ? 

If so, is there a better way to expose it like a potential framebuffer fetch extension ? (I don't think AMD HW supports multiple render targets so things could get sketchy)

**Comment by Degerz** [comment]
How would you like to proceed with this ? 

Can we at least get a vendor extension from AMD exposing this directly if your team doesn't like how interlocks are specified or would you prefer to close this issue if you have no intention of exposing similar functionality in Vulkan ?

I am requesting this because there's arguably a stronger case to have ROV-like functionality exposed in Vulkan rather than in D3D12 because there's a higher interest from open source projects in using it than AAA game engine developers.

**Comment by oscarbg** [comment]
@Degerz @jinjianrong it's sad to see AMD has no plans to support this extension on AMD Vulkan driver even now that VK_EXT_fragment_shader_interlock is a de facto "standard" by the fact that is supported by all other vendors (NV &Intel) and all other OSes (Windows & Linux):
you can see on Windows is supported on "recent" NV  (>=Maxwell) & Intel GPUs (>=gen9 Skylake):
https://vulkan.gpuinfo.org/listdevices.php?platform=windows&extension=VK_EXT_fragment_shader_interlock
similarly on Linux supported on NV&Intel:
https://vulkan.gpuinfo.org/listdevices.php?platform=linux&extension=VK_EXT_fragment_shader_interlock

heck even since Metal2.0 on MacOS we have support for exact same feature.. on Vega cards on Imac Pro we get "RasterOrderGroupsSupported"..

adding more use cases: 
should be useful for VKD3D project for supp. D3D12 "rasterizer ordered views" in case any D3D12 games use it..

well in fact, Xenia emulator D3D12 backend uses ROV feature for better emulation of Xbox EDRAM hardware.. also Xenia is working towards adding Linux support:
https://github.com/xenia-project/xenia/issues/1430
and has a more immature Vulkan backend that Linux backend would use..
Xenia VK backend could take advantage of VK_EXT_fragment_shader_interlock for better/faster emulation of Xbox hardware..
so joining @Triang3l to discussion in case it wants to discuss further..

**Comment by Degerz** [comment]
@oscarbg Good idea to get more people interested in this functionality, I think I'll do the same as well! While you're at it, can you go request other AMD engineers like @Anteru on twitter to show the that the community wants this functionality as well on AMD HW on Vulkan. 

cc @tadanokojin @hrydgard @pent0 

The above have actively expressed interest and/or already using functionality similar to shader interlock in their projects. One of their main motivations to using Vulkan is getting access to modern GPU features like interlock so we'd prefer it from AMD if we didn't have to move over to platform specific APIs like Metal or D3D12 to be able to use this feature! 

Vulkan subpasses are possibly not powerful enough for their purposes. I don't care if AMD doesn't ever expose `VK_EXT_fragment_shader_interlock` but please at least give another viable alternative for their sake even if it is an AMD specific extension!

**Comment by pent0** [comment]
I just want to do programmable blending. If you guys can provide another primitives it would also be ok. Texture barrier is what I am using but not the fastest path really. I dont really know how you guys would do it though

**Comment by ryao** [comment]
> How would you like to proceed with this ?

Would this extension also be unsupported on Windows? If it is, here is my suggestion. Use it. Tell people using Windows to either buy Nvidia hardware or switch to Linux because AMD refuses to support the extension and cite this issue. Tell people using Linux to switch to RADV and cite this issue. Watch AMD backpedal on this very quickly after an executive hears about the situation.

**Comment by Triang3l** [comment]
From the passes point of view, what's different in this from regular image/buffer stores?

**Comment by Degerz** [comment]
@ryao Seems like an unlikely scenario that it would reach to the very high echelons in the company and I don't know if mesa developers are all that interested since I haven't seen any patches related to this issue ... 

@Triang3l Here are some [insights](https://www.saschawillems.de/blog/2018/07/19/vulkan-input-attachments-and-sub-passes/) from Sascha. Along with the stated limitations, I do not think that Vulkan subpasses are capable of handling self-intersecting draws just [like](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_texture_barrier.txt) OpenGL's texture barrier.

**Comment by RussianNeuroMancer** [comment]
> Seems like an unlikely scenario that it would reach to the very high echelons in the company

News article on Phoronix could help with this a bit.

**Comment by ryao** [comment]
> @ryao Seems like an unlikely scenario that it would reach to the very high echelons in the company and I don't know if mesa developers are all that interested since I haven't seen any patches related to this issue ...

All that you need is for it to end users to start telling each other that AMD graphics hardware is not friendly to emulators. It will reach the upper echelons when they are trying to figure out why they did not meet their sales projections.

As for the mesa developers, they might not know that this extension has any use cases. I was under the impression that those working on RADV were volunteers, so if you don’t ask them about it, they seem less likely to implement it.

**Comment by Degerz** [comment]
@ryao TBH, I feel it is more constructive for developers like @pent0 to just express their desire to expose this feature and just list out their use cases instead ... 

At the end of the day, advanced system emulation doesn't even account for the fraction of AMD's customers and the emulation community is already aware that AMD has a checkered history with them so the leading hardware vendor is already favoured over there. 

I'd prefer it if we can show that their driver manager's position is out of touch with the community's position because unlike with higher ups such as executives there's no guarantee that they'd understand this issue or that they'd be specialists regarding GPU drivers to help us out.

**Comment by pent0** [comment]
I love you guys. I know you can do it however the hard it's. Go go go! We all want this feature.

Also programmable blending is not what emulators want also, it's also what many game developers desired to achieve nice and godly effect on their game, which fixed blending can not do. I can't bring an example for PC but here is an example one doing on [Metal IOS](https://medium.com/@heinapurola/engine-internals-optimizing-our-renderer-for-metal-and-ios-77aeff5faba).

Programmable shader pipeline has been here for 15 years, so programmable blending should be too. Its the defacto nowadays.(I copy this quote from this [article](https://forum.beyond3d.com/threads/why-isnt-programmable-blending-more-widely-supported-history-and-discussion.60087/)).

I am really bad at wordings hehe, I just express what many people want. Reconsider please :)

**Comment by ReaperOfSouls1909** [comment]
AMD would be an amazing company if only you stopped making horrible choices and drivers what a shame

**Comment by ryao** [comment]
VKD3D does Direct3D 12 -> Vulkan translation. It will need this extension to implement the related portion of Direct3D 12. That will likely be needed by Valve’s Proton compatibility layer for running Windows software on Linux in the future. Game console emulators are not alone in this matter.

**Comment by jarrard** [comment]
Who exactly does this affect anyway? just PowerVR GPU users?  if so I can understand why AMD doesn't want to dedicated valuable development time to this endeavour. Nothing is stopping the community adding it themselves thanks to **OPEN-SOURCE** drivers!

**Comment by pent0** [comment]
Are you retarded? If you really don't know what @Degerz is talking about
then please shut up and don't interfere.

We are talking about fragment shader interlock, and an example of it is is
to emulate a feature of PowerVR GPU. The main topic is fragment shader
interlock, not PowerVR GPU.

> Nothing is stopping the community adding it themselves thanks to
*OPEN-SOURCE* drivers!

Yes, but assuming anyone outside AMD can implement it.

Vào 12:35, CN, 25 Th8, 2019 jarrard <notifications@github.com> đã viết:

**Comment by jarrard** [comment]
> PowerVR GPU

Then that was a not the best example, it would have been better to give examples that are not fringe case but a more common use case. 
Also referring to people as retarded is why stuff like this gets ignored, its quite a anti-open-source attitude to have, and only derails things.

**Comment by pent0** [comment]
Well, its only me who saying you are retard. Cause the topic is literally
in title. Dont get butthurted, maybe im really anti open source, but you
are stupid. Read the situation before start typing next time.

Vào 13:14, CN, 25 Th8, 2019 jarrard <notifications@github.com> đã viết:

**Comment by pent0** [comment]
You dont even understand. Read! Its not even relevant to the emulation, its related to everything. DXVK needs interlock to implement ROV, games developer need it to do programmable blending. 

It affects many things  hence this extension exist. Please think more. To not spill more this conversation, i will not talk more. You can consider i lose this conversation between you and me.

**Comment by jfdhuiz** [comment]
@pent0 I get that you are angry. If you feel misunderstood, express that feeling. Make your points and cut out the strong language. Strong language won't help your cause (for innocent bystanders it looks like you're grasping at straws), and it is disrespectful. Your point is much, much stronger without the strong language.

**Comment by pent0** [comment]
Really sorry for the bother, I wasnt control my anger. I will delete my
original comments and make it more clear

Vào 15:06, CN, 25 Th8, 2019 jfdhuiz <notifications@github.com> đã viết:

**Comment by jinjianrong** [comment]
@Degerz Here is the feedback from our Vulkan team regarding the extension:

Whilst we could potentially support this feature, we don't see any use of this functionality in typical applications using our drivers (mostly desktop gaming). This functionality is exposed via DirectX (ROVs) and sees no real use outside of a handful of demo applications.

Additionally, this is an inefficient method of performing the typical thing it's often advocated for - order independent transparency. For such an effect we would usually recommend using one of the many two-pass OIT algorithms out there, and making use of multiple subpasses, with the second pass doing the resolve. This is likely the most portably efficient mechanism you can use that works between desktop and mobile parts. We're thus not inclined to support it, as we'd rather not promote an inefficient technology.

However, if you're looking to do direct emulation, we are not sure that really helps you - perhaps you could elaborate on what it is you're trying to emulate exactly and we may be able to advise on an alternative method?

**Comment by MojoJojoDojo** [comment]
> @Degerz Here is the feedback from our Vulkan team regarding the extension:
> 
> Whilst we could potentially support this feature, we don't see any use of this functionality in typical applications using our drivers (mostly desktop gaming). This functionality is exposed via DirectX (ROVs) and sees no real use outside of a handful of of demos applications

It's being used in Just Cause 3.

**Comment by illusion98** [comment]
Quote from @oscarbg 
[Xenia](https://github.com/xenia-project/xenia) (Xbox 360 Emulator) D3D12 backend uses ROV feature for better emulation of Xbox EDRAM hardware.

**Comment by pent0** [comment]
I don't know much about these stuffs, so I will let the guys who know discuss. I will try to get a workaround.

Hi, in our case, we are trying to emulate a feature from the PowerVR GPU. It's that you can fetch the last fragment data of a texel in color buffer and use it for blending inside the fragment shader. It's like blending but not fixed, but inside the shader (programmable).

For OpenGL on AMD, we are using texture barrier. On Vulkan I'm not sure if that's available (the only thing I know is pipeline barrier so far, but I will look more). What would you advice me to do in this case?

**Comment by Degerz** [comment]
@jinjianrong Thank you very much for the response! 

Support for interlocks/ROVs aren't that compelling in D3D because engine developers are more interested in targeting higher hardware compatibility than using the latest features like I mentioned. By comparison, there are already open-source desktop(!) OpenGL applications out there that are already using interlocks or framebuffer fetch and we would like to be able to target both both Windows and Linux on AMD's Vulkan drivers. 

Also, we don't want to implement order independent transparency with Vulkan subpasses. We want to have the same capability to do **_programmable blending_** for emulation purposes and for this reason alone Vulkan subpasses are not a powerful enough mechanism for this purpose since it possibly(?) can't handle self-intersecting draws like we see with texture barrier. I understand from the hardware people's point of view that primitive ordered pixel shaders can place an ordering constraint when executing fragment shaders and that certainly has undesirable effects in terms of increased latency due to the stalling it causes. 

This feature helps us to emulate systems that have non-standard fixed function blending pipelines and systems that are capable of programmable blending as well via shader framebuffer fetch. The biggest reason why interlocks/ROVs/fetch have an advantage over Vulkan subpasses are because the latter does not cover the edge case of self-intersecting geometry and thus gives in our case incorrectly rendered content!

**Comment by hrydgard** [comment]
As another data point as the author of PPSSPP, the popular Sony PSP emulator, the PSP also has a few blend modes that cannot be replicated without fragment shader interlock or similar programmable blending functionality. Now, games don't actually use them much and don't generally use them for self-overlapping draws, so framebuffer copies work in practice to emulate them, but for fully hardware-accurate emulation this would be useful.

I'm not directly involved with Xenia, have only followed its development from the side, but it needs this functionality to simulate some framebuffer formats that only exist on the Xbox 360 and are heavily used by games. They're not practically feasible to emulate in other ways,

**Comment by Triang3l** [comment]
@jinjianrong Xenia needs this for pretty much everything in render target emulation — blending with piecewise linear gamma, with float7e3.7e3.7e3.unorm2 color format, with exponent-biased 16-bit snorm format (with -32 to 32 range); float20e4 depth (especially important when games do EDRAM–RAM–EDRAM round trips (GTA IV, Halo 4) and in this case 32-bit floats cannot be used as games reupload the depth buffer to the EDRAM by writing 24-bit depth to GBA of a R8G8B8A8 color render target aliasing the depth/stencil buffer totally destroying invariance), MSAA via ForcedSampleCount (without that on GPUs without SV_StencilRef we can't restore the stencil buffer after aliasing, and that happens a lot, and have to fall back to slow SSAA).

**Comment by gharland** [comment]
The unordered variant of this extension is essential for voxelization/global illumination/volumetric rendering.  Otherwise what option is there for avg or max blending voxels other than clunky atomiccompswap spinlocks?  Couldn't we at least have the unordered variant?  Even if there were no use case why can't developers just have another tool in the tool box for coming up with new algorithms?

The extension is also requested here, please come over and register your interest.

https://community.amd.com/message/2927066

https://community.amd.com/message/2926956

**Comment by ryao** [comment]
> Its not only relevant to the emulation, its related to everything. DXVK needs interlock to implement ROV, games developer need it to do programmable blending.
> 
> It affects many things hence this extension exist. Please think more.

DXVK does D3D11, which does not need this. VKD3D does D3D12, which does need it to support software using it.

**Comment by Degerz** [comment]
@jinjianrong I'm sure you've realized it by now but our stance on this issue as a community is non-negotiable so we do not desire to seek the 'alternatives' that you speak of. 

I understand the anxiety your team is facing right now since they're going to expose an unfriendly hardware feature and if you absolutely cannot have the general public wanting to access this feature then I have a solution which is applying **_whitelists_** to these community projects specifically to be able to access this feature in the driver.

Is whitelisting certain applications a viable solution at your end for our case ?

**Comment by Triang3l** [comment]
@Degerz Whitelisting will make adding this to new projects impossible, it would never exist in Xenia if we had to go any procedure of being added to a whitelist (and ROV usage there began as an experiment anyway).

**Comment by Degerz** [comment]
@Triang3l Then what other ways do you suggest to an unwilling driver team ? 

If new projects just pop up, then they should arguably just file an appeal since AMD does not like the way applications could potentially use this feature.

**Comment by Triang3l** [comment]
@Degerz To modify the extension so it's more clear about subpass dependencies, if needed — Vulkan extensions are versioned, as long as no new "must"s are added, it should be fine. Whitelisting is completely orthogonal to the actual issue.

@jinjianrong What are the exact issues that interlocking causes with subpasses? I don't know Vulkan much, I've mostly used D3D12, but is it an implicit dependency across subpasses (shader executions for different subpasses running in parallel may still interlock when they don't need to)?

Well, I don't think it should be a big issue. Since it seems to be a hardware limitation, then probably ¯\_(ツ)_/¯ for now, and if more granular interlocking control is added in the future, it could be exposed via a device feature flag + a new `pNext` in `VkRenderPassCreateInfo`?

Or what is incorrect in the extension? But if something is just a hardware limitation making things a bit less optimal, it should be okay to implement it anyway.

**Comment by Degerz** [comment]
@Triang3l I am not here to play around with ideological purity because AMD engineers here are very clearly spooked by anyone using this feature with implications of high performance pitfalls. 

What I want to do is to strive to find a common ground between us which is the community and their driver team and if this means making compromises like whitelists to be able to even access the feature at all then you should at the very least consider this approach as well. 

I assume that AMD will be reasonable enough to grant a whitelist to certain applications as long as the user(s) who filed an issue can explain why this feature fixes their issue for their application and if they absolutely need it.

**Comment by Triang3l** [comment]
@Degerz This is just a graphics programming GitHub issue tracker, calm down a bit :)

Whitelisting is an ideological workaround (that would go against the idea of open drivers and also add unnecessary management burden), we need to settle on something technical. APIs have always been defined in part by what is already there in production GPUs, with tiers in D3D and Vulkan being very explicit about things like device features, a "tier 1" implementation based on the current extension would be enough to expose the existing hardware feature and to cover all the current usage cases, but for flexibility, it may be expanded later.

It's not prohibitively slow, everything has its own performance implications and developers are more or less aware of them.

**Comment by Degerz** [comment]
@Triang3l 

> every tool has its own performance implications and developers are more or less aware of them

Wrong assumption to make that developers are aware of them. There's a good reason why AMD doesn't expose push descriptors despite being a Khronos promoted extension and that's because they're dirt slow on AMD hardware and without AMD engineers reeducating the developers to do the 'alternatives' then the performance of those applications would've been a disaster on their hardware. 

You clearly haven't considered the driver team's feeling on this issue. Taking a hardline stance on this matter is shortsighted when we're at the absolute mercy of their team's decision. I feel if we have to give them a sense of security for them to come to a compromise then whitelists become a reality no matter how much you dislike them.

**Comment by gharland** [comment]
@jinjiangrong , could you suggest an alternative to this without interlock?

This is the kind of code we have to write on AMD to blend RGBA8 voxel fragments.  With interlock it's a simple load modify store inside a critical section.

void imageAtomicAverageRGBA8(layout(r32ui) coherent volatile uimage3D voxels, ivec3 coord, vec3 nextVec3)
{
    uint nextUint = packUnorm4x8(vec4(nextVec3,1.0f/255.0f));
    uint prevUint = 0;
    uint currUint;
 
    vec4 currVec4;
 
    vec3 average;
    uint count;
 
    //"Spin" while threads are trying to change the voxel
    while((currUint = imageAtomicCompSwap(voxels, coord, prevUint, nextUint)) != prevUint)
    {
        prevUint = currUint;                    //store packed rgb average and count
        currVec4 = unpackUnorm4x8(currUint);    //unpack stored rgb average and count
 
        average =      currVec4.rgb;        //extract rgb average
        count   = uint(currVec4.a*255.0f);  //extract count
 
        //Compute the running average
        average = (average*count + nextVec3) / (count+1);
 
        //Pack new average and incremented count back into a uint
        nextUint = packUnorm4x8(vec4(average, (count+1)/255.0f));
    }
}

**Comment by Degerz** [comment]
The irony of it all will be lost on some people but it's a tragedy how this community wants a feature that AMD has identified to be a slow-path which is exactly what they were hoping to avoid with OpenGL and their minefields of slow-paths since quite a few developers couldn't be trusted to hit the fast paths. 

If the AMD driver team have nightmares about OpenGL and are afraid of Vulkan turning into the same disaster then please consider exercising some extra precaution like whitelists or vendor specific extensions! I want to make it as easy as possible for them in this difficult situation.

**Comment by ryao** [comment]
> @Degerz Here is the feedback from our Vulkan team regarding the extension:
> 
> Whilst we could potentially support this feature, we don't see any use of this functionality in typical applications using our drivers (mostly desktop gaming). This functionality is exposed via DirectX (ROVs) and sees no real use outside of a handful of demo applications.

@jinjianrong Is your Vulkan team’s solution for developers that need this for them to switch to Direct3D? It sounds like their options are to use Direct3D just for AMD hardware on Windows or drop AMD hardware support. My vote is for the latter. Nobody should be forced to implement a Direct3D render path because of AMD hardware.

**Comment by jinjianrong** [comment]
@Degerz and others, thanks for the feedback on the use cases. I will pass this on to our team for further discussion

**Comment by Joshua-Ashton** [comment]
> This functionality is exposed via DirectX (ROVs) and sees no real use outside of a handful of demo applications.

Literally not true, Just Cause 3 uses them in D3D11.3

**Comment by Degerz** [comment]
@jinjianrong Please give us an update of the results of the discussion and thank for hearing out our deliberations! 

Regardless of the decision you and your team makes, I will come to an understanding.

**Comment by ryao** [comment]
@Degerz If AMD does not implement this extension, please consider my suggestion of pooling resources with others affected by this on a Vulkan -> Direct3D 12 translation layer just for AMD hardware. Also, @Joshua-Ashton is the developer doing Direct3D 9 -> Vulkan who has talked about doing Vulkan -> Direct3D 12 as a future project, so you are in good company there. :)

**Comment by Anteru** [comment]
ROVs have seen use in games, but it’s extremely limited. Even worse, if you just use ROVs to do blending, performance is typically much worse than alternative solutions, as fundamentally they require a serialization/locking on every single pixel. Due to those reasons, ROVs are of low priority for us to implement. We understand that this priority setting is not super-helpful for your specific case where you’re trying to emulate some specific bit of hardware, but we need to weight it with all the other requests we’re seeing and how many titles will benefit.

If it’s just blending your interested in, you might want to look at techniques like per-pixel linked lists which allow arbitrary programmable blending (with good performance, assuming you have only a few layers), or 64-bit atomics (which allow comparing color + depth simultaneously.)

As mentioned previously, AMDVLK is an open source project, and we’d be happy to review patches if you really want to have it implemented, but there’s not enough bandwidth on our side at this point in time to commit to this feature. Should things change, we’ll let you know.

**Comment by Leopard1907** [comment]
I like how it turned from " that is not the right formula" to " not a pripority, you can implement on your own if you want ".

As seen from the years AMD always offers subpar solutions compared to Nvidia solutions on Khronos based api's.

That being said ; Nvidia is dominant anyway and they will always. So people choose their gpu's regardless. For the case of Xenia ; 90 percent of users will be on an Nvidia gpu anyway. ( Both Linux and Windows )

So i would say Xenia just have to keep using ROV at Vulkan backend. People with AMD on Windows could be able to keep using D3D12 backend on Windows. For Linux AMD users ; nearly everybody is on Radv driver. So it is likely it will be implemented it in Radv by Valve.

I can't see a problem here.

Just typical AMD moves.

**Comment by pent0** [comment]
Hi, thanks for answering. I understand that this is weighted, I respect that tbh. But you guys do consider to provide and working on it when more attentions and titles get to it, so I hope that will be it someday, though I don't think emulators weight much here though. Thanks anyway.

For our case it's hard to do the per-pixel linked list (yes i just search it a couple of minutes ago). We are translating shaders and have no control over it, so we don't know where is the subroutine that does the blending, and where is the subroutine that does the color calculation. It's up to shader writer honestly, so we can't collect all layer's colors and start doing blending all together (we dont even have control over where the blending is).

Generally though, for normal developers, I think per-pixel linked lists are good performance (that trade over memory). **But in our case**, I think some like me will sacrifice performance for accuracy and the ability to be useable if possible.

I hope one day it will be the issue that you guys implement hehe. I dont know anything tbh about drivers and how they do it, but we still want it if possible. Thanks for responding though, I will wait and try to find other workaround in mean time.

**Comment by Anteru** [comment]
It won't hurt for sure :) That said, AMDVLK on Windows uses a different compiler, so some work in that area will remain.

**Comment by oscarbg** [comment]
@Joshua-Ashton  Just Cause 3 uses them in D3D11.3 =anyway this extension would be useful for DXVK then, right?

**Comment by Triang3l** [comment]
@Anteru How much work was needed for ROV support, just for reference? :) Though ROV semantics are less compiler-friendly than shader interlock semantics probably.

**Comment by MojoJojoDojo** [comment]
Well seeing as things won't change here maybe someone could open an issue onmesa-aco github page and ask them to implement it? 
Seeing as it's actually used in a game sold on Steam called Just Cause 3.
This is the repo if someone is interested : https://github.com/daniel-schuermann/mesa/issues

**Comment by Joshua-Ashton** [comment]
D3D11.3 does not imply ROV

**Comment by pent0** [comment]
@Anteru Sorry for the ping, can I ask if fragment shader ordering is possible to consider?

**Comment by pent0** [comment]
Thanks for the info.

Vào 08:56, T.3, 27 Th8, 2019 Degerz <notifications@github.com> đã viết:

**Comment by pent0** [comment]
Performance is definitely be worse, since ordering is waiting for last render to finish. 

But we would want it to work then nothing at all. As I know so far from my side there is really no easy solution to this other than interlock (we are translating shader, and in a scene in a game, multiple fragment shaders use different blending methods and its hard to control over them).

Though thanks, AMD seems worse on this part. I dont really have another idea (degerz advises me with those stuffs too, but it seems there is no alternative solution he can provide this time). I will just use the extension for now and dont care about which gpu support it.

**Comment by pent0** [comment]
Me and other graphics dev discuss and decide to ignore the non exsistence of this extension this time, as there is no other workaround from our side. Users can still use OpenGL backend though, but not really fast.

I have same wish like degerz, I hope the hardware part will improve enough, maybe one day programmable blending will be standard.

**Comment by gharland** [comment]
@Anteru So if it's the primitive ordering that's the problem, the unordered aspect of the extension would still be useful and performant for the pixel_interlock_unordered part of the spec, which is all we want for voxels/volumetric/GI:

"This extension can be useful for algorithms that need to access per-pixel
    data structures via shader loads and stores.  Such algorithms using this
    extension can access such data structures in the critical section without
    worrying about other invocations for the same pixel accessing the data
    structures concurrently."

and the inefficiency lies with the serialization due to primitive ordering:

"Additionally, the ordering guarantees are useful
    for cases where the API ordering of fragments is meaningful.  For example,
    applications may be able to execute programmable blending operations in
    the fragment shader, where the destination buffer is read via image loads
    and the final value is written via image stores."

could you please consider implementing just the unordered variant to begin with?VK_AMD_pixel_interlock_unordered??

**Comment by phire** [comment]
> I don't know what you guys are doing in your emulators, but you may want to consider non-ROV alternatives.

Basically, we need (want, I guess) bit-prefect emulation of exotic alpha blending equations.

We want to read the old color in, blend it with the new color with arbitrary pixel shader code and then write the result out. No sorting, no managing data structures is memory. Just read, modify, write.

I'm not sure how relevant performance tests of OIT with/without ROV is to this use case.

-------------

> How about a per-pixel linked list for overlapping primitives? Store off primitive id and final color, then sort and blend afterward. Basically how (I think) OIT works, but not OIT. Sorry if that's a dumb idea.

I guess linked lists is technically feasible, but it's nowhere near as simple as OIT examples.   
OIT examples use a single blending equation over the entire frame, so they can run a single second pass at the end, while an emulated game might continuously switch between many different blend modes within a single frame. 

You would either need to do a resolve pass every single time the blend mode changed, or write the blend mode to the linked list.

And then it's very common for games on these old console to randomly start rendering with depth writes disabled. We would need to detect this and force a linked list resolve before each draw without depth tests enabled. Or we would need to put these depth writes into the same linked list.....

That's just a few examples of curve-balls that I can think up.

Trying to design a linked list implementation that works with every single edge case an emulated game might throw at us is a nightmare. In comparison, fragment shader interlock is dead simple to implement and even if it has a large speed penalty it might still be the best solution.

**Comment by pent0** [comment]
Correct

**Comment by phire** [comment]
> @jpark37 Per-pixel linked lists still don't give us the needed primitive ordering I think ?

Oh right, I missed that. 

You would need to store a draw call id and primitive id into the linked list and sort by that during resolve.

**Comment by phire** [comment]
@Degerz we always had the option of submitting just one primitive per draw call to emulate programmable blending. It's just a question of performance. 

Like I said, I guess per-pixel linked lists would work, I've just never really considered it before.

But the worst case would have us spamming that resolve pass so many times per frame that I have no idea what the performance of such an approach is like.

**Comment by pent0** [comment]
It work, but pouring the effort that imo is too complex and too hard that near impossiblr is not worth it.

**Comment by pent0** [comment]
Basically, if we want to do it, for case of vita3k, give each draw call an id, store it in linked list, detect all subroutines that do blending in each shader used in a frame, generate a shader that do blending together, reorder blending code with corresponding call, in just one frame. This is like even slower then ROV in our case

**Comment by pent0** [comment]
[Persona 4: Dancing All Night (JP) - Heartbeat, Heartbreak (Video & Let's Dance)](https://www.youtube.com/watch?v=qzJv6O40yUM)

I still have my hope though

**Comment by phire** [comment]
> It would be cool if we could see the Dolphin emulator team implement both approaches

Sounds like you are trying to bait me into doing your homework.

**Comment by MojoJojoDojo** [comment]
I know that Redream uses per-pixel linked lists for Dreamcast emulation if someone wonders which emulator used this approach before.

**Comment by phire** [comment]
@Degerz You know, I've just realized you are trying to implement a 5 series PowerVR gpu, which already has fully programmable blending via a framebuffer fetch capability. 

To actually emulate that with per-pixel linked lists you have to put the shaders entire outstanding state into the linked list, and then somehow resume execution from the resolve shader. 

To resume execution, you either have to generate a resolve shader which inlines the tail of every single programmable blend shader it might encounter or write a resolve shader that executes arbitrary bytecode out of uniforms. The latter is not as crazy as it might seem, Dolphin's ubershaders are essentially interpreters that interpret TEV shader opcodes in the pixel shader.  

fragment interlock would be a much more sane solution for that.

As for Dolphin, the only thing we can't do actually do with the fixed function blend unit is dithered blends into 6bit framebuffers, for example:

`DST_COLOR = ((DST_COLOR<<2 * DST_ALPHA<<2 + SRC_COLOR * (255 - DST_ALPHA<<2)) + dither_table[screen_x_pos & 3, screen_y_pos & 3]) >> 2`

The effect on the final image is usually so minor that it's hardly worth doing.

I'm currently experimenting with hacking per-pixel linked lists into dolphin, because you nerd sniped me.

**Comment by pent0** [comment]
Its SGX 5xx gpus. And one degerz mentioned is from ps vita.

> To resume execution, you either have to generate a resolve shader which inlines the tail of every single programmable blend shader it might encounter or write a resolve shader that executes arbitrary bytecode out of uniforms. The latter is not as crazy as it might seem, Dolphin's ubershaders are essentially interpreters that interpret TEV shader opcodes in the pixel shader.

Its not 200x console anymore. These shaders on Vita are complex than that. Which p4g maybe an example, laid programmable blending across multiple opcodes, so we need to detect dependency. Also its insane that you need to order them correctly cause multiple shaders use multiple blending method in a scene. I think you said that before. Essentially this is like from high efficient method turned into bottleneck, because we have to generate shader every frame, and shader may also does complex operations, which make keeping the framerate stable not possible.

If I want it to work, I just want to take an easier path. This is like insane work, i dont like it.

**Comment by pent0** [comment]
Overall, interlock may sacrifies some perf, but its easiest method and everyone likes it. Its also used for compute.

On AMD, the performance sacrifition maybe bigger then other vendor, so we cant take the easy 0ath. We may need to wait for hardware to improve or more traffic came to this. I will take the easy part.

**Comment by phire** [comment]
@inolen It's a shame, dolphin looks like it is just over the size of a vec4. 

32bits of color/alpha, 24bits of depth.

I think we can squash the blend state into 30 bits:  
3 bits src factor  
3 bits src factor alpha  
3 bits dst factor  
3 bits src factor alpha  
4 bits of logic ops mode  
6 misc control bits  
8 bits of constant alpha  

So unless we can squash the next index and primitive index into just 42 bits total, it's not going to fit.

**Comment by phire** [comment]
@jpark37 Correct, that's what I'm suggesting in my edit

@Degerz My understanding is that shader cores (nvidia, amd and intel) are totally capable doing indirect branches and have been able to for years. It's just not really exposed to pixel shaders. OpenGL 4.0 does have ARB_shader_subroutine extention, but that only exposes uniform control flow that is static per draw call.

The downside is the whole wave/wrap follows the branch. To do dynamic per thread indirect branching, you would have to disable lanes and loop, executing upto 64 different subroutines before continuing. Performance would be very dependant on how often threads branch to the same blend shader.

But I think GCN and later is technically capable of this. I wonder if Nvidia added hardware to Tesla to accelerate this operation?

Also, I think you would be surprised at the performance of literally writing a bytecode interpeter in your pixel shader and compiling blend programs to bytecode which you store in uniforms. Keep the bytecode simple, only the operations you need and only four or eight registers, spill any extras to the stack. Your resolve shader will just use a switch statement and bunch of dynamic indirect array accesses to execute it.

It won't be as fast as the option above, but potentially competitive, within a order of magnitude.   
More importantly, it doesn't require the development of any special extensions and works across many GPUs (though watch out for dynamic array indexing performance on Nvidia, they don't have an "index into array of registers" operation like most other vendors)

**Comment by oscarbg** [comment]
Just two additional cents:
@Degerz if ROV use cases on Vulkan end using VK_KHR_shader_atomic_int64 then we will have to ask Intel Windows Vulkan driver team to implement it (or use interlock EXT on this case).. available everywhere else on Linux and Win..

in the light of test where AMD use of ROV causes a 20X slowdown wondering if AMD big slowdowns are really to architecture not being so friendly to ROV as NV and Intel or really they have to additionaly implement to software workarounds to HW bugs in ROV case than cause more slowdowns than would be needed either way.. I may be wrong but one hint may be Xenia D3D12 ROV has notable graphical issues running Read Redemption on Vega the two last times I tested (and there ware 6 good months between).. maybe @Triang3l can share thoughts on AMD rendering bugs on Xenia D3D12 ROV on Vega at least.. and if anybody knows if new Navi 5700 cards are known to render correctly on Xenia D3D12 ROV case..

**Comment by oscarbg** [comment]
hey guys.. good news to share: we will have interlock ext supported on MacOS MoltenVK very soon:

https://github.com/KhronosGroup/SPIRV-Cross/pull/1138

so even AMD GPUs on MacOS (Vega only right now) will support it..
maybe interesting if DXVK or VKD3D gained support for ROV to run the Intel demos mentioned earlier to see if also the big slowdowns on AMD GPUs stil are there..

**Comment by Degerz** [comment]
Also, do we tell [them](https://github.com/citra-emu/citra/pull/3778) to just sort the linked lists by primitive id and store the shadow map framebuffer combiner state too as a solution ?

Attempting a bytecode interpreter for our case to generate blend shader programs could be something to consider in the future once system configurations with GPUs capable of 10+ TFlops/500+ GB/s become common enough which won't be too far off ...

**Comment by MadByteDE** [comment]
> Its not only relevant to the emulation, its related to everything. DXVK needs interlock to implement ROV, games developer need it to do programmable blending.

Hey guys, I'm just a regular consumer and would like to know if this stuff you're discussing here may be the cause for [this kind of artifacting seen on Navi / Raven Ridge ](https://imgur.com/a/3DOWATz) in games utilizing DXVK. Seems like the dev's not gonna answer this any time soon.. This could prevent new ppl from creating more issues for this [over](https://github.com/GPUOpen-Drivers/AMDVLK/issues/105) - [and over](https://github.com/GPUOpen-Drivers/AMDVLK/issues/110) - [again](https://github.com/GPUOpen-Drivers/AMDVLK/issues/106)..

**Comment by Joshua-Ashton** [comment]
@MadByteDE No.

**Comment by mcoffin** [comment]
@MadByteDE nope (I think) this is just about how to expose some new ngg capabilities

**Comment by ReaperOfSouls1909** [comment]
The emulator PCSX2 also plans on using this so lots want it and it be good to have it

**Comment by amayra** [comment]
I guess my next GPU is Nvidia then ?

**Comment by jarrard** [comment]
I'm sure somebody will step up and produce some code towards making it happen. AMD drivers are open source afterall. Nvidia on the other hand, no chance.

**Comment by Joshua-Ashton** [comment]
@jarrard I really don't think you're going to find anyone who would want to spend time adding it to AMDVLK that isn't at AMD.

**Comment by jarrard** [comment]
maybe not AMDVLK, but possibly RADV driver

**Comment by oddMLan** [comment]
@jinjianrong
> Our stance is that we don’t want to implement it. It messes with subpasses and is not the right formulation

WTF? `GL_INTEL_fragment_shader_ordering` worked perfectly fine in the 17.x Radeon drivers. Actually it's a requirement for certain emulation software. You're telling me you don't give us any alternative for fragment interlocking for OpenGL?. Seems that you only care about propietary APIs (DirectX). Your OpenGL drivers suck **big** time. Literally everyone agrees on that point.

https://github.com/PCSX2/pcsx2/wiki/OpenGL-and-AMD-GPUs---All-you-need-to-know
https://dolphin-emu.org/blog/2013/09/26/dolphin-emulator-and-opengl-drivers-hall-fameshame/

Maybe if I told everyone I know to not touch AMD GPUs with a 10 ft pole, and see your sales dwindle, you'll start _wanting_ to implement it. Oh yeah but big thanks for a 3% speedup in ceirtain PC games with each 1.2GB driver update, while your OpenGL support remains terribly pitiful.

**Comment by amayra** [comment]
Friendship ended with _ayymd_
Now **NVIDIA** is my best friend

**Comment by SaltyBet** [comment]
As a heads up, official AMD ROV support _might_ be coming (I'm guessing post-Navi):

[AMD ROV Patent US20200202815A1](https://github.com/GPUOpen-Drivers/AMDVLK/files/5025192/AMD.ROV.Patent.US20200202815A1.pdf)

**Comment by Joshua-Ashton** [comment]
@SaltyBet That was filed in 2018.

**Comment by RinMaru** [comment]
yea its not coming. already got allot of Emu devs looking for alternatives in fear of pissing off AMD users

**Comment by gharland** [comment]
I only skim read it, correct me if I'm wrong, but it looks like a software solution that would be just as slow as a roll-your-own.  

A fast non-order-dependant critical section would still be nice.

**Comment by RinMaru** [comment]
> I only skim read it, correct me if I'm wrong, but it looks like a software solution that would be just as slow as a roll-your-own.
> 
> A fast non-order-dependant critical section would still be nice.

That would be Per-Pixel-Linked OIT iirc its been done on Redream and other DC Emus as of recently to basically work around the issue

**Comment by Joshua-Ashton** [comment]
Why is this closed? The issue is still not resolved.

**Comment by RinMaru** [comment]
> Why is this closed? The issue is still not resolved.

Because AMD isnt going to resolve it its a feature that is hardly used outside the Emulation community. Some devs are looking at other less faster ways to do this because they afraid to piss off the AMD Users

**Comment by Joshua-Ashton** [comment]
You could work around it for programmable blending if the resource is in GENERAL layout (ie. no DCC) and you emit a readback barrier and sample the current framebuffer's image as a normal image and blend that way.

**Comment by Triang3l** [comment]
So I assume the only option we have on Vulkan is a "mutex buffer" with an R32_UINT per pixel + atomic CAS spinlock, though I'm not sure if that would preserve the order of polygons, especially translucent ones with programmable blending (for opaque, apart from manual depth testing, a "primitive index" buffer could possibly be used, rejecting if new primitive index <= last written index, but with blending there's no 1:1 association between a pixel and a primitive/draw anymore, also not sure how wrapping could be handled).

**Comment by RinMaru** [comment]
Per Pixel-Linked is what Redream does for DC's OIT and yea big memory overhead the higher the internal res is.

**Comment by Triang3l** [comment]
@JacobHeAMD @jinjianrong Anyway arbitrarily dropping features that are "not recommended because of being inefficient" (poor architectural compatibility with subpasses is not an unresolvable issue considering Intel and Nvidia are implementing this feature in Vulkan fine) sets a bad precedent and contributes to stagnation of the API (and graphics as a whole) and even more divergence between APIs and issues for those who want to support them all.

Any tool may be used well, just like any tool may be used badly, but it still has its uses. When we choose the tool to do the job, we know the goals and requirements of the task, the limitations that we can agree upon, and we evaluate the advantages of potential solutions and the drawbacks of each.

Let's take the debatable topic of antialiasing as an example. MSAA offers a sharp, stable (no "defocusing" every time a small movement happens) image, preserving small details, and also provides some transparency effects through sample masking, so it's very nice for vivid, sandbox-feeling, immersive games, and pretty much the only option for VR. However, it's noticeably more expensive, for this reason relatively rarely used in this generation (because of consoles, and because it's a bit complicated to integrate into a rendering pipeline with post-processing effects that use depth), and does not help with shading aliasing. Would those be good reasons to drop MSAA completely in the API even though your GPU supports it fine? Leaving developers with two choices — either blurry options eating small details (TAA, less blurry, but jaggy in motion — FXAA, MLAA, SMAA), or falling back to true supersampling, which would solve the goal, but result in something even worse than what you wanted by removing MSAA, because that would have much higher performance costs. But if we have some milliseconds to spare on our target hardware, or can simplify some art, and not that many high-frequency low-roughness objects so specular aliasing is not a big issue (or can selectively supersample some parts of shaders or the frame, that cause the most aliasing, but not everything) — why not? MSAA would solve our goal perfectly considering our requirements and limitations. (I really hope you never ever treat this paragraph as a suggestion… at least you're still implementing EQAA even and are not advocating for reconstruction techniques like DLSS, so I guess here we're pretty safe.)

ROV is also a tool with good and bad sides, but those are factors to consider when using, not a reason to outright make the concept unusable. Yes, it has a flaw of being able to interlock only within individual pixels/samples and thus not letting users benefit from or recreate optimizations used in the fixed-function output-merger like color compression and sample deduplication. But we are aware of that, everything has limitations, rasterization has them, ray tracing has them. It's still a powerful and a valid option for various tasks.

If you're doing programmable blending or order-independent transparency, you don't have to use ROV for every single translucent effect in your frame — you can sort on a coarse level, and only use ROV for fine sorting within certain objects that need that, maybe through an additional framebuffer with premultiplied alpha (possibly with even lower bandwidth and memory usage than with per-pixel linked lists, and without potential overflow in case of large overdraw).

If, like in the emulation case, you're using it for pixel packing — first of all, it's the only solution (apart from TBR-like subpass input, which, however, turns multisampling into supersampling) that allows for maximum accuracy (and unlike, maybe, art fidelity, it's not just some subjective "looks good to me" thing that allows tradeoffs, either the original visuals are reproduced correctly, or it's just broken, in some games less, in some games totally). You may also have plenty of milliseconds to use if you're emulating some 2000s console on a 2020 system, so performance may just not be an issue for you (remember cycle-accurate bsnes and higan also). Actually in Xenia, the naïve purely ROV-based output path (even for depth/stencil) is significantly faster than the traditional render target-based one, since the latter involves a lot of copying to support reinterpreting EDRAM data. However, even with ROV, there's still space for optimization. A conservative host depth buffer may be added so true early Z can work (unlike discarding based on whether the manual depth test passed in the whole quad via ddx/ddy). Another possible optimization is using the fixed-function output-merger where suitable, and only using ROV for formats that don't exist on the PC or for parts of the frame requiring unusual blending — that's not uncommon practice in GPU design too, for instance, where you have fallbacks for cases when cmask/htile may become unusuable.

"Inefficient" is not absolutely measurable. It may be relatively less efficient in some cases, relatively more efficient in other, and it depends on what kind of efficiency you need in each individual case — efficient as in short frame time, or as in completely solving the problem, or even as in development time (Doom Eternal with its 500 FPS is more of an exception than a rule). And you don't need perfect efficiency in all cases, you just need the solution to be efficient enough for your requirements. ROV is efficient enough for us (including on AMD hardware where it's still faster than RTV/DSV in our case), and could work just fine. But now, instead of having it work just fine in our planned Vulkan version, we'll likely have to put a warning asking users to switch to the Direct3D 12 renderer, or to RADV if ROV support is present there, if they're using an AMD GPU. A situation with no winners. At least it's not some S3TC patent status that crippled the functionality for no good, but still there's nothing positive that comes out of simply removing useful tools.

**Comment by Passingby1** [comment]
@jinjianrong @JacobHeAMD 
It's ok, really, our stance is that the green camp is a better all around choice for PC users that want to use their machines in whatever way they see fit.

Awesome job, never change.

**Comment by plonk264** [comment]
@jinjianrong @JacobHeAMD  
please don't "just close it" without responding to these well thought out and written responses by Triang3l, Degerz, and ryao

**Comment by chrismile** [comment]
Just in case someone is interested: I tested the performance of fragment shader interlock (i.e., the OpenGL/Vulkan counterpart of ROVs) for order-independent transparency on NVIDIA hardware: https://chrismile.net/blog/2020/mlab-sync-comp/

TL;DR: In the tests I performed, using ordered fragment shader interlock for Multi-Layer Alpha Blending (MLAB) on NVIDIA hardware was 4% faster than using spinlocks. Furthermore, fragment shader interlock and ROVs can guarantee memory access ordering, while spinlocks can't. Using per-pixel linked lists for alpha compositing was significantly slower than MLAB with fragment shader interlock and has an unbounded memory requirement. I couldn't test the performance on AMD hardware (at least without first having to fully rewrite my program to use Direct3D), as fragment shader interlock is unfortunately not supported.

Furthermore, the authors of Moment-Based Order-Independent Transparency (MBOIT) used ROVs in their DirectX reference implementation of their rendering technique and got even better performance with ROVs than when using hardware-accelerated blending. A quote from the supplementary material of their paper (https://cg.cs.uni-bonn.de/en/publications/paper-details/muenstermann2018-mboit/): "Surprisingly, the resulting frame times on our test hardware are actually shorter than those obtained with hardware-accelerated additive blending. We do not have a conclusive explanation for this phenomenon but note that a low overhead from using rasterizer ordered views is expected since the shader program is very short."

I hope that maybe the decision to not support fragment shader interlock will be reconsidered at some point in the future. It would be a really handy feature to have, even if it were slightly less performant on AMD hardware than on Intel and NVIDIA hardware.

**Comment by OtavioRaposo** [comment]
Just don't use AMD gpus.

**Comment by Doctorwho1909** [comment]
It should be added and the devs who want to use it should be in control of optimizing it

**Comment by jarrard** [comment]
> Just don't use AMD gpus.

Not really sensible solution since AMD is the only Open-Source driver gpu for Linux, except for Intel whom has yet to release their GPU(s). Does Intel support this in their open-source drivers yet? if so then perhaps someone can hack that over to AMD's driver :)

**Comment by v-fox** [comment]
> > Just don't use AMD gpus.
> 
> Not really sensible solution since AMD is the only Open-Source driver gpu for Linux, except for Intel whom has yet to release their GPU(s). Does Intel support this in their open-source drivers yet? if so then perhaps someone can hack that over to AMD's driver :)

Well, it [actually does](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/989).

**Comment by diego-rbb-93** [comment]
@jinjianrong still no feedback to all the emu dev users asking around? :/

**Comment by jarrard** [comment]
Shrug, 
If I was desperate for this feature I'd be looking at what intel is doing to support it then trying to slap that into the radv driver. 
But it seems 100% of the people wanting this feature have absolutely no coding or open-source experience, unfortunately.

**Comment by gharland** [comment]
Waste of time, Jarrard.  Needs to be cross-platform.

**Comment by devshgraphicsprogramming** [comment]
> 
> 
> @Degerz Here is the feedback from our Vulkan team regarding the extension:
> 
> Whilst we could potentially support this feature, we don't see any use of this functionality in typical applications using our drivers (mostly desktop gaming). This functionality is exposed via DirectX (ROVs) and sees no real use outside of a handful of demo applications.
> 
> Additionally, this is an inefficient method of performing the typical thing it's often advocated for - order independent transparency. For such an effect we would usually recommend using one of the many two-pass OIT algorithms out there, and making use of multiple subpasses, with the second pass doing the resolve. This is likely the most portably efficient mechanism you can use that works between desktop and mobile parts. We're thus not inclined to support it, as we'd rather not promote an inefficient technology.
> 
> However, if you're looking to do direct emulation, we are not sure that really helps you - perhaps you could elaborate on what it is you're trying to emulate exactly and we may be able to advise on an alternative method?

Except for Total War : Three Kingdoms

The funny thing is that we're porting a game to ChromeOS to run as an android application, inside the ANGLE sandbox prison, and even GLES 3.1 implemented by ANGLE reports `NV_fragment_shader_interlock` 

And seems AMD is a special boy.

**Comment by marekolsak** [comment]
We could add this into AMD's Mesa GL driver, or we would accept a 3rd-party contribution adding this feature there, at least to the extent of what DX supports.

**Comment by Triang3l** [comment]
What's happening on RDNA 3 with POPS, by the way, with `src_pops_exiting_wave_id` and `HW_REG_POPS_PACKER` being removed? Are you not scheduling overlapping wavefronts until the overlapped ones have completed execution now?

**Comment by ryao** [comment]
> We could add this into AMD's Mesa GL driver, or we would accept a 3rd-party contribution adding this feature there, at least to the extent of what DX supports.

In hindsight, that would be better than nothing.

**Comment by Triang3l** [comment]
@ryao I'm currently researching this (being an ISV and a wannabe contributor (froghacker specifically 🐸), not an AMD engineer), most unanswered questions currently are on the register setup side and things like potentially needed implicit barriers between changes related to multisampling and VRS, though I can't promise anything.

My current plan on the shader side (there on GCN5/RDNA/RDNA2 it consists of two parts — overlapped wave awaiting, and then a loop running the critical section code for each overlap layer within the current wave, effectively splitting a part of the shader into smaller "subgroups") so far is:

1. Find the locations that dominate all begins and post-dominate all ends.
2. Maybe narrow this region to memory accesses.
3. Move the critical region boundaries out of all outer loops, to handle cases when static domination doesn't imply dynamic (like a loop with `if (i == 0)` { beginInvocationInterlock(); } else if (i == 1) { endInvocationInterlock(); }`), and so that there are no `break`s from the original shader that need to be routed from the new loop to the outer loop.
4. Expand the critical section to include all lane-aware operations such as ballots, so that they're unaffected by the new loop, which is not exposed in any way in the SPIR-V control flow, narrowing the exec mask.

**Comment by Triang3l** [comment]
On GFX9 (Vega), I'm getting horrible hangs if `POPS_OVERLAP_NUM_SAMPLES` doesn't match the (rasterizer?) sample count, by the way (haven't checked yet with sample-rate shading, however), so it looks like the per-pixel mode is the only supported there (though per-sample still must be exposed for compatibility with apps relying on it, more ordering is always okay — in an extreme example, a Vulkan implementation processing 1 quad at a time in order is technically valid too). I'm not sure if the glitches I'm getting in [Nvidia's order-independent transparency sample](https://github.com/nvpro-samples/vk_order_independent_transparency) with the per-pixel mode in my experiments are something actually related to the interlock or something else (in the sample itself, or in RADV), the look of the edges is not stable in time, but the effect doesn't look anything like a race condition between waves (no rectangular patterns), I'm also getting some weirdness in the spinlock mode so I guess it's not related to fragment shader interlock. On more recent hardware, primarily RDNA 2 with VRS, I don't know yet, but the registers don't seem like they imply flexibility in the interlock modes at all, apparently the hardware is designed only for the D3D's (and Metal's possibly?) requirement of per-fine-pixel interlocking (though I haven't checked yet whether it's implemented as per-coarse-pixel on RDNA 2 or actually just per-fine-pixel).

![Early test of a custom RADV POPS implementation on the Vega](https://user-images.githubusercontent.com/1764343/197360319-c63cfcb1-c137-4fae-a947-577293b297e7.png)

**Comment by Triang3l** [comment]
Since [it's Halloween](https://twitter.com/Triang3l/status/1586755635131154432), I need to say that the way it's not the right formulation, especially when it comes to how operations aware of which lanes are currently active (like ballots) interact with the intrawave collision loop (specifically, intrawave collisions result in a part of the shader within one wave being executed first for overlapped lanes, then for overlapping lanes, then for other overlapping lanes, and so on — each time with a narrower set of active lanes than outside the CS), is extremely **SCARY** and spooky 🙀😿 For example:

```glsl
uint64_t before = ballotARB();
// let's say `before` is …0000111111111111
beginInvocationInterlockARB();
uint64_t during = ballotARB();
// first iteration: `during` = …0000000000001111
// second iteration: `during` = …0000111111110000
endInvocationInterlockARB();
uint64 after = ballotARB();
// `after` is …0000111111111111 again
```

This is clearly not right from the point of view of GLSL and SPIR-V (or in an even more horrifying example, from the point of view of ROV loads/stores in HLSL) — there are no control flow constructs (conditionals, loops, returns) in the shader code that would suggest that `during` should be different than `before` and `after` — they're in the same control flow tree node, there are no returns between them. Interlocking is exposed purely on the level of a single invocation, which is not true in the implementation on AMD.

Yesterday I was thinking how they can be handled more or less safely, but it looks like for that, it would be necessary to locate all dependency chains of every ballot that cross the boundaries of the critical section, and include them in the CS. However, this has at least two issues.

One inconvenience is that by including a dependency chain of a ballot crossing the CS boundaries into the CS, you're expanding the CS, thus changing its boundaries — and some ballots that could stay outside previously now may have to be moved into it. Though this can be solved by just running this pass again and again until it makes no more changes.

But a more severe problem, that I already explained in my previous message, but that I have to highlight the painfulness of, is that obtaining the dependency chain of something is highly non-trivial when variables (including dynamically indexed arrays), or what's even worse, global memory, are involved. Basically, if you're writing into a buffer or an image, potentially any buffer/image load (with `restrict`, for that specific resorce, but without `restrict`, for any non-`restrict` resource of a type with which aliasing is possible) preceded by that store may be a dependency on that store. And you don't even have to actually write the result of a ballot for the dependency to appear: that would be an extremely weird and rare use case (even more like just a test case), however, there's a much more realistic situation — any write to a manually non-uniform-indexed (for which ballots and first lane reads are used commonly) image/buffer is also a memory dependency on the ballot.

One potential solution that I thought about was forcing all ballots to be inside the critical section. But there's an obvious flaw in it, so I'm of course not going to use it — that would outright ruin the last ballot in the shader that was outside the CS in the original code. Specifically, that would change:

```glsl
criticalSection {
  // ROV accesses here
}
uint64_t lanesRemaining = ballotARB();
while (lanesRemaining) {
  // some non-uniform resource access scalarized manually here
}
```

into:

```glsl
uint64_t lanesRemaining;
criticalSection {
  // ROV accesses here
  lanesRemaining = ballotARB();
}
while (lanesRemaining) {
  // some non-uniform resource access scalarized manually here
}
```

thus `lanesRemaining` would belong not to the whole subgroup, but just to the uppermost overlap layer — but later it would be used for providing data for the entire subgroup. Again, for this to work correctly, all the dependencies of `lanesRemaining` will have to be moved into the critical section.

I'm really not sure that I want to spend a huge amount of time trying to untangle all this mess, so at least at first I'll probably just leave a // FIXME for now. While this would of course result in behavior that makes no sense from the SPIR-V or GLSL point of view, at least the change of the set of active lanes would happen in locations that are predictable and can be taken into account. Specifically, they will be the `OpBeginInvocationInterlockEXT` and `OpEndInvocationInterlockEXT` themselves, or, if they are in different control flow nodes, and thus the boundaries of the critical section need to be moved to a parent control flow node to include both, the critical section will start right before and/or end right after a control flow construct, such as a conditional or a loop — where the writer of the shader (unless they explicitly assume that the condition is uniform) would expect the active lanes to naturally potentially change. For this reason, I'll also drop my idea of narrowing the CS to the memory accesses inside it (which would help if the shader, for instance, is written in GLSL and thus adheres to the extremely strict control flow rules of GL_ARB_fragment_shader_interlock, and puts the CS begin/end instructions on the outermost control flow level, but does all the memory accesses between them conditionally), as that would essentially result in something even more broken, unpredictable and uncontrollable — basically in the way ballots interact with ROV accesses on Direct3D.

Of course we could use more radical solutions, such as going Intel's `sendc` way and disabling intrawave collisions completely if ballots are used if that's possible on the hardware (loading intrawave collisions is switchable separately from loading overlapped wave info for some reason, but I'm not sure if disabling that reliably works) — or if not, wrapping the whole shader in the CS (but then it may be reasonable to switch off `EXEC_IF_OVERLAPPED` too, but again, I don't know if that can be relied on). However, this edge case of interaction of ballots and FSI is probably too weird for real usage — but what's much more realistic is that a shader would want the critical section to be as short as possible, especially if it does just programmable blending (like some simple overlay/hard light) somewhere in the end, it wouldn't make much sense not to do some heavy, but independent, lighting work earlier in the shader in parallel between overlapped and overlapping invocations, no matter if they are in the same wave or in different. So in the end, I think I'll just keep ballots broken, but broken in a controllable way, as the alternatives that I can imagine seem to be unreasonable.

**Comment by Triang3l** [comment]
@jpark37 While this was very long ago, if you still remember the details of your [tests](https://github.com/GPUOpen-Drivers/AMDVLK/issues/108#issuecomment-525144229), could you please provide the settings you had in the Intel OIT sample?

Most importantly, was MSAA used in your test run on AMD, and what exact algorithm was used without ROV in your testing setup (if any OIT at all)? MSAA specifically has a massive performance hit with ROV on AMD due to adjacent primitives overlapping each other as I found out two comments above, but that applies only to the PixelInterlock modes. Without MSAA, or with MSAA in the SampleInterlock mode, in Nvidia's OIT sample, in a spinlock -> interlock comparison, I was getting a ratio similar to your Intel and Nvidia results (on the RX Vega 10, without MSAA, 22ms > 26ms if I recall correctly). MSAA with PixelInterlock, on the other hand, was closer to what you were getting on AMD, though even worse — a 15x-ish increase. However, the spinlock is also an approach that's very hostile to parallelism, so maybe the spinlock was just slow in the first place, and the interlock turned out to be just slightly slower.

Though I'll probably also try running it by myself when I finish other tasks. I also wonder, by the way, since it's a D3D11 sample, whether implicit UAV barriers might have caused a significant drop, or was ROV actually the bottleneck there.

**Comment by jpark37** [comment]
> While this was very long ago, if you still remember the details of your https://github.com/GPUOpen-Drivers/AMDVLK/issues/108#issuecomment-525144229, could you please provide the settings you had in the Intel OIT sample?

Sorry, I don't.

> Most importantly, was MSAA used in your test run on AMD

This much is unlikely though because I'll always disable MSAA if the option is in front of me.

**Comment by devshgraphicsprogramming** [comment]
> > Why is this closed? The issue is still not resolved.
> 
> Because AMD isnt going to resolve it its a feature that is hardly used outside the Emulation community. Some devs are looking at other less faster ways to do this because they afraid to piss off the AMD Users

3 AAA games use it, AMD has this feature on DX12.

Also we plan to use it to do CSG in a single pass for a CAD app, at this rate we'll open a popup saying "Buy a real GPU" and open a browser with Amazon and Ebay searches for Nvidia and Intel when we detect an AMD GPU.

**Comment by Triang3l** [comment]
Yes, and I did more research recently for my future blog post — and the deterministic ordering, and thus the lack of temporal noise, in overflow handling makes fragment shader interlock a much more reliable solution even to order-independent transparency compared to other two-pass methods like a spinlock. This was also cited in the [GRID 2 article](https://www.intel.com/content/dam/develop/external/us/en/documents/grid-2-case-study-revised-393278.pdf) (for them even 2 nodes per pixel were enough for order-independent transparency, and 4 nodes for Adaptive Volumetric Shadow Maps for smoke lighting), in the [MLAB benchmark](https://chrismile.net/blog/2020/mlab-sync-comp/). Without fragment shader interlock, no matter how advanced your tail blending algorithm is, it will _always be broken_ anyway — because you'll just have incoherent noise if _any_ overflow happens. Like an analog TV with no antenna connected, on your trees or in your glass panes.

Additionally, with fragment shader interlock you can do OIT partially, coarsely sorting large batches of geometry (level map tiles, objects, meshlets), and doing fine OIT inside those batches and between nearby batches, including to handle intersecting polygons (which are very common in foliage). And if you sort batches by the farthest depth in them (conservatively is enough), with fragment shader interlock, you can compare the sort key of the current batch with the closest OIT fragment depth in the pixel so far — and if it turns out to be closer, with fragment shader interlock, you can just safely resolve OIT as soon as that happens and free all your OIT layers for reuse (for pixels that actually need that only, unlike with a pipeline barrier — which also wouldn't work with instancing or mesh shading) as you know that all new fragments will be closer from that moment on. This can effectively provide you infinite layers in the view, with a small number of layers within object "clusters" needed in the RAM.

Other uses that come to mind are deferred decals — blending the normal G-buffer, as [demonstrated by Nvidia](https://docs.nvidia.com/gameworks/content/gameworkslibrary/graphicssamples/opengl_samples/normalblendeddecalsample.htm) (especially useful for decals on curved surfaces); or drawing huge numbers of sorted particles with a custom blending equation (like Hard Light for both lightening and darkening), as well as with per-particle blending equation selection (especially useful with bindless textures — to have all the additive fireworks and all the alpha-blended smoke in a single draw command with correct ordering between each other).

On the implementation side, by the way, I'm somewhat worried about the changes to POPS setup introduced by GFX11. Specifically, the POPS_OVERLAP_NUM_SAMPLES setting is now gone. Currently it's difficult for me to allocate the money to purchase a testing device, so I can't check this by myself. But can someone (@Anteru possibly?) please confirm, how does POPS behave on RDNA 3 with sample-rate shading?

Direct3D and Metal (and Intel's old extension) only require sample-granularity interlocking with sample-frequency shading. However, Vulkan and OpenGL fragment shader interlock gives explicit control of the interlock scope to the shader via its execution mode — so it's still possible to request pixel-level interlock in a sample-rate shader (like via POPS_OVERLAP_NUM_SAMPLES = 0 on the Vega/RDNA/RDNA2). And if the device supports the `fragmentShaderPixelInterlock` feature, it should expose whole-pixel-scope interlock _regardless of the shading frequency_ of the shader. However, without the `fragmentShaderPixelInterlock` feature, it's _not possible to support ROVs on DXVK and VKD3D_ at all, as `PixelInterlock` execution modes can't be used even conditionally, while they would be required for multisampling without sample-rate shading — even though in native Direct3D ROVs would just work naturally.

**Comment by Triang3l** [comment]
@TheLastRar Fragment shader interlock and shader framebuffer fetch (not an explicit barrier that causes excessive synchronization and doesn't support draw commands with overlap, let alone intersecting primitives, but Arm's ordered version) both can be used to implement programmable blending, however, they're quite different in the details, so it would be ideal if hardware supported both, but with Intel being the only PC graphics card developer supporting the latter currently (while all the biggest 3 have fragment shader interlock in their hardware and at least some of their drivers), FSI is effectively the only option on the PC now.

But in general, fragment shader interlock offers **massively** more flexibility than shader framebuffer fetch.

The only advantage of SFBF I can imagine is that it supports late depth/stencil test, so it works directly with things like alphatested surfaces. With fragment shader interlock, the write happens in the shader, so your only choices are early depth/stencil (with post-depth coverage with MSAA), which only works for opaque surfaces not modifying the depth or the stencil reference from the shader, or full-blown software depth testing.

However, FSI, being a shader part rather than an output-merger one, allows for arbitrary addressing, and that removes lots of limitations:

- The layout and the amount of pixel-local data can be managed completely freely. SFBF restricts you to (R32G32B32A32 * maximum output attachment count) of data at most, and if you use more than one attachment, your data becomes scattered across multiple image subresources. With FSI, you can store arbitrary amounts of data with any layout you want.
- Random access within pixel-local data. Relative addressing of temporary registers is a painful thing, on AMD for instance the offset is placed in a scalar register, not a vector one. But more importantly, you don't always need to access _all_ the per-pixel data. With SFBF, if you want to modify data at some dynamic location, you need to fetch all the data, and then to export it back. An example of when this may happen is order-independent transparency — if there's no overflow, you don't need to tail-blend and thus to locate the least important fragment to replace throughout all the pixel, you only need to increment the counter and to add the new fragment, so there's no need to waste the bandwidth reading and writing the entire list. Another example would be single-pass construction of trees, like for the Z axis when doing 2D-tiled lighting with conservative rasterization — again, you only need to visit the nodes that matter for the new light.
- Not only the inner layout, but the global placement of the pixel-local data is controlled by the application. The data doesn't have to be placed in a rectangle in one or multiple image subresources. This is important for emulation, including Xenia in particular, where we emulate EDRAM addressing of the console, thus data reinterpretation (which happens a lot on the console, even for clears that are done via 4x MSAA depth-only rectangle draws there) between different framebuffer offsets, widths, MSAA sample counts, can be done with no copying — just a pipeline barrier is sufficient.

Also I'm not entirely sure about the requirements of Arm's rasterizer-ordered attachment access (maybe subpassInputMS is allowed, but I don't know for sure), but compared to OpenGL ES SFBF, FSI has one advantage for MSAA — you still can use pixel-frequency shading with FSI, and access per-sample data based on the input coverage mask with sample interlock, or per-pixel data as usual with pixel interlock. The OpenGL ES SFBF specification says "Reading the value of gl_LastFragData produces a different result for each sample. This implies that all or part of the shader be run once for each sample…", but in reality as far as I know it's always full fallback to sample-rate shading, which cancels out the idea of MSAA on the performance side.

Note that with FSI, you can still take advantage of texture tiling (if I understand correctly, it's even the same 64KB_R_X for both framebuffer attachments and storage images on RDNA and RDNA 2 normally), and as far as I know, modern AMD GPUs support internal compression for storage images as well.

But again, what's the most important is that there's no SFBF anywhere on the PC except for Intel GPUs (and maybe Innosilicon, Moore Threads, though I don't know the details about them).

**Comment by devshgraphicsprogramming** [comment]
> umbers of sorted particles with a custom blending equation (like Hard Light for both lightening and darkening), as well as with per-particle blending equation selection (especially useful with bindless textures — to have all the additive fireworks and all the alpha-blended smoke in a single draw command with correct ordering between each other — and you can't just put fireworks in one draw command and smoke in another, as

Another fun use for ROV is rendering to and blending non-renderable formats like RGB9E5 or some custom stuff.

**Comment by devshgraphicsprogramming** [comment]
> You briefly mentioned custom blending equations, and many have also mentioned wanting this extension for programmable blending, it's possible to use feedback loops for that purpose https://registry.khronos.org/vulkan/specs/1.2-extensions/html/vkspec.html#renderpass-feedbackloop
> 
> Mesa zink uses this to implement fbfetch (https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/12603), which Mesa uses to support the `GL_KHR_blend_equation_advanced` extension.
> 
> PCSX2 uses it to support non-standard blend modes. This requires splitting render passes and using barriers (with VK_DEPENDENCY_BY_REGION) to ensure sync. Performance is reasonable on Nvidia (even with a very large number of draws), however, AMD doesn't fully support VK_DEPENDENCY_BY_REGION and has worse performance as a result, see this [Reddit post](https://www.reddit.com/r/emulation/comments/rz7rqj/pcsx2_vulkan_released_in_latest_dev_builds/hrwqmbb/?context=3) & [AMD Community post](https://community.amd.com/t5/opengl-vulkan/vulkan-poor-performance-due-to-barrier-region-bit-being-ignored/td-p/501962), which seems to be a hardware limitation.
> 
> I don't know if OIT can be supported by the above approach. and I don't know how performance compares to using shader interlocks w.r.t. custom blending.

IIRC unless you have the brand new Vulkan ARM/EXT externsion meant to replace OpenGL ES Pixel Local Storage or FramebufferFetch.... subpass feedback loops are only limited to a single pixel overwrite cycle, then you need a barrier (can be by-region).

PLS + PSI are best of both worlds, because you can use the local framebuffer/tiler memory to store your MLAB4 buckets and not a `coherent storageImage`.

**Comment by Triang3l** [comment]
I guess this is not an April Fools' Day joke? 😜 Even though this situation is complete tragicomedy and farce 🤷‍♂️
https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22250

**Comment by John-Gee** [comment]
> I guess this is not an April Fools' Day joke? stuck_out_tongue_winking_eye Even though this situation is complete tragicomedy and farce man_shrugging [gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22250](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22250)

Fantastic news, thank you for your work on this!

**Comment by rijnhard** [comment]
It just got merged into Mesa RADV devel

**Comment by r2rX** [comment]
Congratulations, @Triang3l, and well done. AMD users are indebted....only appreciation and admiration for this awesome contribution. :)

**Comment by oddMLan** [comment]
Brilliant! The FOSS community has done it again 🎉
I guess this goes to show support on Windows is possible, right? 
...........
What do you mean there aren't any open source drivers on Windows?
...........
What do you mean AMD has to implement it? But they just said they won't! ☹️

https://www.youtube.com/watch?v=Lo4DMz6fZG0

**Comment by Moonlacer** [comment]
Hello AMD Vulkan developers! Has your stance on supporting this extension changed within the last 4 years? There seems to be plenty of examples given here on how this would affect the user experience on any AMD card using the Windows proprietary drivers, so I would really like to know your current (updated) thoughts on this matter.

Best regards, Moonlacer

**Comment by Squall-Leonhart** [comment]
VK_EXT_fragment_shader_interlock has been added to amdvlk in [194a181da7e2cca5f70ec0f9e65119955b3d2b47](https://github.com/GPUOpen-Drivers/xgl/commit/194a181da7e2cca5f70ec0f9e65119955b3d2b47)

**Comment by John-Gee** [comment]
It's the same driver only the compiler may differ.

**Comment by mirh** [comment]
The same commit also added `VK_KHR_maintenance5` (which did in fact land in 23.12.1)
Shader interlock is still nowhere to be seen though (it isn't exposed yet with *availableExtensions.AddExtension*)

**Comment by Triang3l** [comment]
>  VK_EXT_fragment_shader_interlock has been added to amdvlk in [194a181da7e2cca5f70ec0f9e65119955b3d2b47](https://github.com/GPUOpen-Drivers/xgl/commit/194a181da7e2cca5f70ec0f9e65119955b3d2b47)

That commit doesn't implement it, only makes it "known" to the extension management in the driver, and the newly added device feature query reports that all interlock features are unsupported. But it's an awesome sign that it's soon™! Unfortunately no new AMDVLK will be released for the Vega generation though.

**Comment by Squall-Leonhart** [comment]
VK_EXT_fragment_shader_interlock is added as of Adrenaline 24.12.1
